### PR TITLE
Feature/account model

### PR DESCRIPTION
**Pull request by xal9wiii4ik**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -22,3 +22,5 @@ RUN poetry install
 
 FROM base as prod
 RUN poetry install --no-dev
+
+CMD ['flask', 'db', 'upgrade']
```
Comment (artsiom-streetshares) [inline Dockerfile]: I would not do it here. this is like a preparation before deploying the app, not the build instruction. your build will fail if there's no database. this should be invoked by docker-compose.

**Comment by artsiom-streetshares** [comment]
every PR must have meaningful title add a meaningful description. 

sample title: Add user registration/login, user account model and CRUD api

sample description: 
Motivation: As a user I want to be able to be able to have standard capabilities like registration and login. Also to create and update my data.
What have been done: Added account model, CRUD apis, corresponding tests.

File: app/account/models.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    id = model.Column(model.Integer(), primary_key=True, autoincrement=True)
+    username = model.Column(model.String(length=100), unique=True)
+    password = model.Column(model.String(length=500))
+    email = model.Column(model.String(length=100))
+    first_name = model.Column(model.String(length=100))
+    last_name = model.Column(model.String(length=100))
+    is_staff = model.Column(model.BOOLEAN())
+
+    def __init__(self, username: str,
```
Comment (artsiom-streetshares) [inline app/account/models.py]: this is not needed, provided by framework

File: app/account/permissions.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    Args:
+        user: current user model
+        request: current request
+        pk: current user pk
+    Return:
+        has access or no
+    """
+
+    if request.method in ['PATCH', 'DELETE']:
+        return bool(pk == user.id)
```
Comment (artsiom-streetshares) [inline app/account/permissions.py:18]: ``
```suggestion
        return pk == user.id
```

File: app/account/models.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+from main import db as model
+
+
+class Account(model.Model):
+    """
+    Model for table account
+    """
+
+    __tablename__ = 'account'
```
Comment (artsiom-streetshares) [inline app/account/models.py]: this is by default your model name in snake case, so can omit

File: app/account/models.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    id = model.Column(model.Integer(), primary_key=True, autoincrement=True)
+    username = model.Column(model.String(length=100), unique=True)
+    password = model.Column(model.String(length=500))
+    email = model.Column(model.String(length=100))
+    first_name = model.Column(model.String(length=100))
+    last_name = model.Column(model.String(length=100))
+    is_staff = model.Column(model.BOOLEAN())
```
Comment (artsiom-streetshares) [inline app/account/models.py:15]: is BOOLEAN all uppercase indeed?

File: app/account/serializers.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import typing as tp
+
+from account.schemas import AccountData
+
+
+def account_serializer(model_objects: tp.Any,
```
Comment (artsiom-streetshares) [inline app/account/serializers.py:6]: why Any if that's supposed to be for Account?

File: app/account/serializers.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    Returns:
+      dict with user data
+    """
+
+    schema = AccountData
+
+    if not many:
+        if model_objects is not None:
+            return schema(**model_objects.__dict__).__dict__, 200
+        return {}, 404
```
Comment (artsiom-streetshares) [inline app/account/serializers.py]: serializer shouldn't be returning any http status codes, it's job is to turn stuff into json

File: app/account/services_views.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+def create_new_account(data: RegisterUser) -> tp.Dict[str, tp.Union[str, bool]]:
+    """
+    Create new account
+    Args:
+        data: register data
+    Return:
+        dict with account data
+    """
+
+    from main import db
```
Comment (artsiom-streetshares) [inline app/account/services_views.py:34]: why it's here?

File: app/account/services_views.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+    new_account = Account(username=data.username,
+                          password=data.password,
+                          is_staff=False,
+                          email=data.email,
+                          first_name=data.first_name,
+                          last_name=data.last_name)
+    db.session.add(new_account)
+    db.session.commit()
+    account_data = data.__dict__
```
Comment (artsiom-streetshares) [inline app/account/services_views.py:45]: that's weird you use __dict__ ? it's rarely needed to be used directly. pretty sure pydantic can get to it by itself

File: app/account/services_views.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    """
+    Create access token using credentials or refresh token
+    Args:
+        token_type: access or refresh
+        data: user credentials
+    Returns:
+        new token
+    """
+
+    helping_data: tp.Any = {
```
Comment (artsiom-streetshares) [inline app/account/services_views.py:76]: this Any doesn't help

File: app/account/services_views.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+        if token_type != TOKEN_TYPE:
+            return jsonify({'error': 'Invalid token type'})
+
+        try:
+            payload = jwt.decode(access_token, SECRET_KEY, algorithms=[TOKEN_ALGORITHM])
+            if payload['sub'] != 'access':
+                return {'error': 'Access token is expected'}, 401
+            account = Account.query.filter_by(username=payload['username']).first()
+            request.user = account  # type: ignore
```
Comment (artsiom-streetshares) [inline app/account/services_views.py:151]: why you use both user and account words? you can choose one and use it everywhere

File: app/account/services_views.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+def get_user_from_pk(pk: int) -> type_model_data:
+    """
+    Get user from pk
+    Args:
+        pk: current user pk
+    Return:
+        dict with user data
+    """
+
+    from account.models import Account
```
Comment (artsiom-streetshares) [inline app/account/services_views.py:168]: why import here?

File: app/main.py [hunk 13]
```diff
@@ -1,22 +1,70 @@
...
+
+def models_initialization() -> None:
+    """
+    Initialize models
+    Args:
+        db: current db state
+    """
+
+    logging.info('Starting initialization of models')
+    app.app_context().push()
```
Comment (artsiom-streetshares) [inline app/main.py]: I don't think you need to do it manually

File: app/main.py [hunk 14]
```diff
@@ -1,22 +1,70 @@
...
+    db.init_app(app)
+    db.create_all()
```
Comment (artsiom-streetshares) [inline app/main.py]: this should be done by alembic separately from application startup (this is a part of deploy)

File: app/migrations/versions/04a7e7949c24_initial_migration.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+# revision identifiers, used by Alembic.
+revision = '04a7e7949c24'
+down_revision = None
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('account',
```
Comment (artsiom-streetshares) [inline app/migrations/versions/04a7e7949c24_initial_migration.py:21]: where is password?

File: app/migrations/versions/bde1aa0ecc9c_initial_migration.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
+"""Initial migration.
```
Comment (artsiom-streetshares) [inline app/migrations/versions/bde1aa0ecc9c_initial_migration.py:1]: duplicate migration. as a general principle target one migration per PR

File: app/settings.py [hunk 17]
```diff
@@ -9,6 +8,44 @@
...
+
+# DB SETTINGS
+POSTGRES_HOST: str = os.environ.get('POSTGRES_HOST')  # type: ignore
+POSTGRES_USER: str = os.environ.get('POSTGRES_USER')  # type: ignore
+POSTGRES_PASSWORD: str = os.environ.get('POSTGRES_PASSWORD')  # type: ignore
+POSTGRES_DB: str = os.environ.get('POSTGRES_DB')  # type: ignore
+POSTGRES_PORT: str = os.environ.get('POSTGRES_PORT')  # type: ignore
+
+# TEST DB SETTINGS
+TEST_POSTGRES_HOST: str = 'db-test'
```
Comment (artsiom-streetshares) [inline app/settings.py:23]: why you need test db notion? you should just set correct db credentials during tests

File: app/tests/test_account/test_services.py [hunk 18]
```diff
@@ -0,0 +1,35 @@
+from account.services_views import create_tokens, generate_access_token_from_refresh
+
+
+class TestRefreshToken:
```
Comment (artsiom-streetshares) [inline app/tests/test_account/test_services.py]: you don't need the class, just use plain functions

File: docker-compose.yml [hunk 19]
```diff
@@ -19,5 +21,12 @@ services:
     env_file:
       - ./app/.env
 
+  db-test:
```
Comment (artsiom-streetshares) [inline docker-compose.yml:24]: why you need a separate test db?

File: pyproject.toml [hunk 20]
```diff
@@ -12,7 +12,11 @@ python-dotenv = "^0.19.2"
 Flask-SQLAlchemy = "^2.5.1"
 psycopg2-binary = "^2.9.2"
 gunicorn = "^20.1.0"
-pre-commit = "^2.16.0"
+Flask-Migrate = "^3.1.0"
+Flask-Script = "^2.0.6"
```
Comment (artsiom-streetshares) [inline pyproject.toml:16]: why you need this?

File: Dockerfile (hunk 1 above)
Comment (xal9wiii4ik) [reply Dockerfile]: Done

File: app/account/models.py (hunk 2 above)
Comment (xal9wiii4ik) [reply app/account/models.py]: done

File: app/account/permissions.py (hunk 3 above)
Comment (xal9wiii4ik) [reply app/account/permissions.py:18]: this is helps for mypy

File: app/account/models.py (hunk 4 above)
Comment (xal9wiii4ik) [reply app/account/models.py]: Done

File: app/account/models.py (hunk 5 above)
Comment (xal9wiii4ik) [reply app/account/models.py:15]: i dont know; pycharm told me that it can be

File: app/account/serializers.py (hunk 6 above)
Comment (xal9wiii4ik) [reply app/account/serializers.py:6]: circular import) in next version i removed this

File: app/account/services_views.py (hunk 8 above)
Comment (xal9wiii4ik) [reply app/account/services_views.py:34]: db.commit, in next version i removed this

File: app/account/services_views.py (hunk 9 above)
Comment (xal9wiii4ik) [reply app/account/services_views.py:45]: pydantic return BaseModel with your fields

File: app/account/serializers.py (hunk 7 above)
Comment (xal9wiii4ik) [reply app/account/serializers.py]: done

File: app/account/services_views.py (hunk 11 above)
Comment (xal9wiii4ik) [reply app/account/services_views.py:151]: didnt understand what you were talking about, explain please

File: app/account/services_views.py (hunk 12 above)
Comment (xal9wiii4ik) [reply app/account/services_views.py:168]: meaning in func? circular import

File: app/main.py (hunk 13 above)
Comment (xal9wiii4ik) [reply app/main.py]: done

File: app/main.py (hunk 14 above)
Comment (xal9wiii4ik) [reply app/main.py]: removed

File: app/migrations/versions/04a7e7949c24_initial_migration.py (hunk 15 above)
Comment (xal9wiii4ik) [reply app/migrations/versions/04a7e7949c24_initial_migration.py:21]: this is ask for alembic?)
removed in next version

File: app/migrations/versions/bde1aa0ecc9c_initial_migration.py (hunk 16 above)
Comment (xal9wiii4ik) [reply app/migrations/versions/bde1aa0ecc9c_initial_migration.py:1]: removed in next version

File: app/settings.py (hunk 17 above)
Comment (xal9wiii4ik) [reply app/settings.py:23]: this is for creating new account and etc and drop test data

File: docker-compose.yml (hunk 19 above)
Comment (xal9wiii4ik) [reply docker-compose.yml:24]: this is for creating new account and etc and drop test data

File: app/tests/test_account/test_services.py (hunk 18 above)
Comment (xal9wiii4ik) [reply app/tests/test_account/test_services.py]: done

File: pyproject.toml (hunk 20 above)
Comment (xal9wiii4ik) [reply pyproject.toml:16]: moved to dev dependencies

File: app/account/services_views.py (hunk 10 above)
Comment (xal9wiii4ik) [reply app/account/services_views.py:76]: this is helps for mypy(typing cant see that datetime in tp.Dict[..., datetime], i dont know why)

File: app/account/serializers.py (hunk 6 above)
Comment (artsiom-streetshares) [reply app/account/serializers.py:6]: what next version?

File: app/account/services_views.py [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+    except Exception:
+        return {'error': 'Invalid token or token is expired'}, 401
+
+
+def authenticate(func: tp.Any) -> tp.Any:
+    """
+    Decorator for authenticate user(set current user to request)
+    """
+
+    from flask import request, jsonify
```
Comment (artsiom-streetshares) [inline app/account/services_views.py:131]: something is wrong with your code if you need so many local imports

File: app/account/services_views.py (hunk 11 above)
Comment (artsiom-streetshares) [reply app/account/services_views.py:151]: what is the difference between user and account?

File: app/account/serializers.py (hunk 6 above)
Comment (artsiom-streetshares) [reply app/account/serializers.py:6]: function should almost always be start with the verb

File: app/account/services_views.py (hunk 9 above)
Comment (artsiom-streetshares) [reply app/account/services_views.py:45]: I don't understand what your reply means

Comment (artsiom-streetshares) [reply app/account/services_views.py:45]: as I said you rarely need to access __dict__. pydantic can work with orm models https://pydantic-docs.helpmanual.io/usage/models/#orm-mode-aka-arbitrary-class-instances

File: app/account/views.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+    if status_code == 404:
+        return jsonify(user_data), status_code
+
+    has_permission = account_permission(user=request.user, request=request, pk=pk)  # type: ignore
+
+    if not has_permission:
+        return jsonify({'Error': 'You has not permissions to perform this action'}), 401
+
+    if request.method == 'PATCH':
+        data = AccountData(**request.json).__dict__  # type: ignore
```
Comment (artsiom-streetshares) [inline app/account/views.py:83]: you don't need to access dict

File: app/account/views.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        for key in data.copy().keys():
+            if data[key] == '':
```
Comment (artsiom-streetshares) [inline app/account/views.py:85]: why you need to do this here?

File: app/account/views.py [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+                del data[key]
+        update_account(pk=pk, data=data)
+    elif request.method == 'DELETE':
+        remove_account(pk=pk)
+        return jsonify({}), 204
```
Comment (artsiom-streetshares) [inline app/account/views.py:90]: why return empty object? just return nothing

File: app/migrations/versions/04a7e7949c24_initial_migration.py (hunk 15 above)
Comment (artsiom-streetshares) [reply app/migrations/versions/04a7e7949c24_initial_migration.py:21]: I don't understand your reply. the table doesn't contain account password

File: app/migrations/versions/bde1aa0ecc9c_initial_migration.py (hunk 16 above)
Comment (artsiom-streetshares) [reply app/migrations/versions/bde1aa0ecc9c_initial_migration.py:1]: still not fixed

File: app/migrations/versions/fc466d2ecaaa_update_account_model.py [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+# revision identifiers, used by Alembic.
+revision = 'fc466d2ecaaa'
+down_revision = '04a7e7949c24'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_table('account')
```
Comment (artsiom-streetshares) [inline app/migrations/versions/fc466d2ecaaa_update_account_model.py:21]: why drop table?

**Pull request closed without merging**